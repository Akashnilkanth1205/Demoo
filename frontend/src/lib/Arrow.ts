/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Column, Table } from "apache-arrow"
import { range, unzip } from "lodash"

import { IStyler } from "autogen/proto"

interface Schema {
  index_columns: (string | RangeIndex)[]
  columns: SchemaColumn[]
  column_indexes: any[]
}

interface RangeIndex {
  kind: "range"
  name: string | null
  start: number
  step: number
  stop: number
}

interface SchemaColumn {
  field_name: string
  metadata: Record<string, any> | null
  name: string | null
  numpy_type: string
  pandas_type: string
}

interface TableDimensions {
  headerRows: number
  headerColumns: number
  dataRows: number
  dataColumns: number
  rows: number
  columns: number
}

type TableCellType = "blank" | "index" | "columns" | "data"

interface TableCell {
  type: TableCellType
  id?: string
  classNames: string
  content: string
}

export interface Styler {
  uuid?: string
  caption?: string
  styles?: string
  displayValues?: Arrow
}

export class Arrow {
  private table: Table

  private schema: Schema

  private index: string[][]

  private columns: string[][]

  private data: any[][]

  private styler?: Styler

  constructor(dataBuffer: Uint8Array, styler?: IStyler) {
    this.table = Table.from(dataBuffer)
    this.schema = this.getSchema()
    this.index = this.getIndex()
    this.columns = this.getColumns()
    this.data = this.getData()
    this.styler = {
      uuid: styler?.uuid || undefined,
      caption: styler?.caption || undefined,
      styles: styler?.styles || undefined,
      displayValues: styler?.displayValues
        ? new Arrow(styler.displayValues)
        : undefined,
    }
  }

  private getSchema(): Schema {
    const schema = this.table.schema.metadata.get("pandas")
    if (schema == null) {
      throw new Error("Table schema is missing.")
    }
    return JSON.parse(schema)
  }

  private static getColumnData(column: Column): any[] {
    return range(0, column.length).map(rowIndex => column.get(rowIndex))
  }

  private getIndex(): string[][] {
    return this.schema.index_columns.map(field => {
      const isRangeIndex = typeof field === "object" && field.kind === "range"
      if (isRangeIndex) {
        const { start, stop, step } = field as RangeIndex
        return range(start, stop, step).map(String)
      }
      const column = this.table.getColumn(field as string)
      return Arrow.getColumnData(column).map(String)
    })
  }

  private getColumns(): string[][] {
    const isMultiIndex = this.schema.column_indexes.length > 1
    return unzip(
      this.schema.columns
        .map(column => column.field_name)
        .filter(fieldName => !this.schema.index_columns.includes(fieldName))
        .map(fieldName =>
          isMultiIndex
            ? JSON.parse(
                fieldName
                  .replace(/\(/g, "[")
                  .replace(/\)/g, "]")
                  .replace(/'/g, '"')
              )
            : [fieldName]
        )
    )
  }

  private getData(): any[][] {
    const rows = this.table.length
    const columns = this.columns.length > 0 ? this.columns[0].length : 0

    return range(0, rows).map(rowIndex =>
      range(0, columns).map(columnIndex =>
        this.table
          .getColumnAt(columnIndex)
          ?.get(rowIndex)
          ?.toString()
      )
    )
  }

  get tableId(): string | undefined {
    return this.styler?.uuid && `T_${this.styler.uuid}`
  }

  get tableStyles(): string | undefined {
    return this.styler?.styles
  }

  get caption(): string | undefined {
    return this.styler?.caption
  }

  get dimensions(): TableDimensions {
    const [headerColumns, dataRowsCheck] = this.index.length
      ? [this.index.length, this.index[0].length]
      : [0, 0]

    const [headerRows, dataColumnsCheck] = this.columns.length
      ? [this.columns.length, this.columns[0].length]
      : [0, 0]

    const [dataRows, dataColumns] = this.data.length
      ? [this.data.length, this.data[0].length]
      : // If there is no data, default to the number of header columns.
        [0, dataColumnsCheck]

    if (
      (dataRows !== 0 && dataRows !== dataRowsCheck) ||
      (dataColumns !== 0 && dataColumns !== dataColumnsCheck)
    ) {
      throw new Error(
        "Table dimensions don't align: " +
          `rows(${dataRows} != ${dataRowsCheck}) OR ` +
          `cols(${dataColumns} != ${dataColumnsCheck})`
      )
    }

    const rows = headerRows + dataRows
    const columns = headerColumns + dataColumns

    return {
      headerRows,
      headerColumns,
      dataRows,
      dataColumns,
      rows,
      columns,
    }
  }

  public getCell(rowIndex: number, columnIndex: number): TableCell {
    const { headerRows, headerColumns, rows, columns } = this.dimensions

    if (rowIndex < 0 || rowIndex >= rows) {
      throw new Error("Row index is out of range.")
    }
    if (columnIndex < 0 || columnIndex >= columns) {
      throw new Error("Column index is out of range.")
    }

    const isBlankCell = rowIndex < headerRows && columnIndex < headerColumns
    const isIndexCell = rowIndex >= headerRows && columnIndex < headerColumns
    const isColumnsCell = rowIndex < headerRows && columnIndex >= headerColumns

    if (isBlankCell) {
      const classNames = ["blank"]
      if (columnIndex > 0) {
        classNames.push(`level${rowIndex}`)
      }
      return {
        type: "blank",
        classNames: classNames.join(" "),
        content: "",
      }
    }

    if (isIndexCell) {
      const dataRowIndex = rowIndex - headerRows

      const uuid = this.styler?.uuid
      const classNames = [
        `row_heading`,
        `level${columnIndex}`,
        `row${dataRowIndex}`,
      ]

      return {
        type: "index",
        id: uuid
          ? `T_${uuid}level${columnIndex}_row${dataRowIndex}`
          : undefined,
        classNames: classNames.join(" "),
        // Table index is stored as is (in the column format).
        content: this.index[columnIndex][dataRowIndex],
      }
    }

    if (isColumnsCell) {
      const dataColumnIndex = columnIndex - headerColumns

      const classNames = [
        `col_heading`,
        `level${rowIndex}`,
        `col${dataColumnIndex}`,
      ]

      return {
        type: "columns",
        classNames: classNames.join(" "),
        content: this.columns[rowIndex][dataColumnIndex],
      }
    }

    const dataRowIndex = rowIndex - headerRows
    const dataColumnIndex = columnIndex - headerColumns

    const uuid = this.styler?.uuid
    const classNames = ["data", `row${dataRowIndex}`, `col${dataColumnIndex}`]

    const content = this.styler?.displayValues
      ? this.styler.displayValues.getCell(rowIndex, columnIndex).content
      : this.data[dataRowIndex][dataColumnIndex]

    return {
      type: "data",
      id: uuid
        ? `T_${uuid}row${dataRowIndex}_col${dataColumnIndex}`
        : undefined,
      classNames: classNames.join(" "),
      content,
    }
  }
}
