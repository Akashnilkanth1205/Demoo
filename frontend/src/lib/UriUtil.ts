/**
 * @license
 * Copyright 2018-2019 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


/**
 * host:port tuple
 */
export interface BaseUriParts {
  host: string;
  port: number;
}

/**
 * Create a ws:// or wss:// URI for the given path.
 */
export function buildWsUri({host, port}: BaseUriParts, path: string): string {
  const protocol = isHttps() ? 'wss' : 'ws'
  return `${protocol}://${host}:${port}/${path}`
}

/**
 * Create an HTTP URI for the given path.
 */
export function buildHttpUri({host, port}: BaseUriParts, path: string): string {
  const protocol = isHttps() ? 'https' : 'http'
  return `${protocol}://${host}:${port}/${path}`
}

/**
 * True if we're connected to the host via HTTPS.
 */
function isHttps(): boolean {
  return window.location.href.startsWith('https://')
}
