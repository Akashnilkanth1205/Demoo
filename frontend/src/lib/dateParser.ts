import parser, { Parser, Eat, Tokenizer } from "remark-parse"
import { DateTimeHandler } from "lib/DateTime"
import { Datetime } from "autogen/proto"
function dates() {
  //@ts-ignore
  var Parser = this.Parser
  var tokenizers = Parser.prototype.inlineTokenizers
  var methods = Parser.prototype.inlineMethods

  tokenizers.datetime = tokenizeDatetimes as Tokenizer
  // Run it just before `text`.
  methods.splice(methods.indexOf("text"), 0, "datetime")
}

tokenizeDatetimes.notInLink = true
tokenizeDatetimes.locator = locateDate

function locateDate(value: string, fromIndex: number) {
  return value.indexOf("{@", fromIndex)
}

function tokenizeDatetimes(eat: Eat, value: string, silent: boolean) {
  var match = /{@datetime: (.*)}(.*$)/.exec(value)
  if (match) {
    if (silent) {
      return true
    }

    const date = Datetime.fromObject(JSON.parse(match[1]))
    console.log(JSON.parse(match[1]), date)
    // Reconstruct markdown
    const markdownParts = [
      { type: "inlineCode", value: DateTimeHandler.dateToString(date) },
    ]
    if (match.index) {
      markdownParts.unshift({
        type: "text",
        value: value.substring(0, match.index),
      })
    }
    if (match[2]) {
      markdownParts.push({ type: "text", value: match[2] })
    }

    return eat(value)({
      type: "root",
      children: markdownParts,
    })
  }
}

export default dates
