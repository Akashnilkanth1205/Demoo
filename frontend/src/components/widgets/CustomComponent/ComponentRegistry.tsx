/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BaseUriParts, buildHttpUri } from "lib/UriUtil"

/**
 * Downloads and compiles components from the server.
 */
export class ComponentRegistry {
  private readonly getServerUri: () => BaseUriParts | undefined

  public constructor(getServerUri: () => BaseUriParts | undefined) {
    this.getServerUri = getServerUri
  }

  public getComponentURL(componentId: string, path: string): string {
    const serverURI = this.getServerUri()
    if (serverURI === undefined) {
      throw new Error("Can't fetch component: not connected to a server")
    }

    return buildHttpUri(serverURI, `component/${componentId}/${path}`)
  }
}
