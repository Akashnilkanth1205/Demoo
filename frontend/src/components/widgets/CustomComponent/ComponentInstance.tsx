/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Map as ImmutableMap } from "immutable"
import { logError, logWarning } from "lib/log"
import { Source, WidgetStateManager } from "lib/WidgetStateManager"
import React, { createRef, ReactNode } from "react"
import { ComponentRegistry } from "./ComponentRegistry"

/** Messages from Component -> Streamlit */
export enum ComponentBackMsgType {
  // A component sends this message when it's ready to receive messages
  // from Streamlit. Streamlit won't send any messages until it gets this.
  // No data.
  COMPONENT_READY = "componentReady",

  // The component has a new widget value. Send it back to Streamlit, which
  // will then re-run the app.
  // Data: { value: any }
  SET_WIDGET_VALUE = "setWidgetValue",

  // The component has a new height for its iframe.
  // Data: { height: number }
  SET_FRAME_HEIGHT = "setFrameHeight",
}

/** Messages from Streamlit -> Component */
export enum ComponentForwardMsgType {
  // Sent by Streamlit when the component should re-render.
  // Data: { args: any, disabled: boolean }
  RENDER = "render",
}

interface Props {
  registry: ComponentRegistry
  widgetMgr: WidgetStateManager

  disabled: boolean
  element: ImmutableMap<string, any>
  width: number
}

interface State {
  frameHeight?: number
}

// TODO: catch errors and display them in render()

export class ComponentInstance extends React.PureComponent<Props, State> {
  private iframeRef = createRef<HTMLIFrameElement>()
  // True when we've received the COMPONENT_READY message
  private componentReady = false
  private lastRenderArgs = {}
  private lastRenderDataframes = []

  public constructor(props: Props) {
    super(props)
    this.state = { frameHeight: undefined }
  }

  public componentDidMount = (): void => {
    if (this.iframeRef.current == null) {
      // This should not be possible.
      logError(
        `ComponentInstance does not have an iframeRef, and will not receive messages!`
      )
      return
    }

    if (this.iframeRef.current.contentWindow == null) {
      // Nor should this.
      logError(
        `ComponentInstance iframe does not have an iframeRef, and will not receive messages!`
      )
      return
    }

    this.props.registry.registerListener(
      this.iframeRef.current.contentWindow,
      this.onBackMsg
    )
  }

  public componentWillUnmount = (): void => {
    if (
      this.iframeRef.current == null ||
      this.iframeRef.current.contentWindow == null
    ) {
      return
    }

    this.props.registry.deregisterListener(
      this.iframeRef.current.contentWindow
    )
  }

  /**
   * Receive a ComponentBackMsg from our component iframe.
   */
  private onBackMsg = (type: string, data: any): void => {
    switch (type) {
      case ComponentBackMsgType.COMPONENT_READY:
        // Our component is ready to begin receiving messages. Send off its
        // first render message! It is *not* an error to get multiple
        // COMPONENT_READY messages. This can happen if a component is being
        // served from the webpack dev server, and gets reloaded. We
        // always respond to this message with the most recent render
        // arguments.
        this.componentReady = true
        this.sendForwardMsg(ComponentForwardMsgType.RENDER, {
          args: this.lastRenderArgs,
          dfs: this.lastRenderDataframes,
        })
        break

      case ComponentBackMsgType.SET_WIDGET_VALUE:
        if (!this.componentReady) {
          logWarning(
            `Got ${type} before ${ComponentBackMsgType.COMPONENT_READY}!`
          )
        } else {
          this.handleSetWidgetValue(data, { fromUi: true })
        }
        break

      case ComponentBackMsgType.SET_FRAME_HEIGHT:
        if (!this.componentReady) {
          logWarning(
            `Got ${type} before ${ComponentBackMsgType.COMPONENT_READY}!`
          )
        } else {
          this.handleSetFrameHeight(data)
        }
        break

      default:
        logWarning(`Unrecognized ComponentBackMsgType: ${type}`)
    }
  }

  /** The component set a new widget value. Send it back to Streamlit. */
  private handleSetWidgetValue = (data: any, source: Source): void => {
    const value = tryGetValue(data, "value")
    if (value === undefined) {
      logWarning(`handleSetWidgetValue: missing 'value' prop`)
      return
    }

    const widgetId: string = this.props.element.get("id")

    // TODO: handle debouncing, or expose some debouncing primitives?
    // TODO: ints, arrays, "button triggers", ... dataframes?

    this.props.widgetMgr.setJsonValue(widgetId, value, source)
  }

  /** The component has a new height. We'll resize the iframe. */
  private handleSetFrameHeight = (data: any): void => {
    const height: number | undefined = tryGetValue(data, "height")
    if (height === undefined) {
      logWarning(`handleSetFrameHeight: missing 'height' prop`)
      return
    }

    this.setState({ frameHeight: height })
  }

  private sendForwardMsg = (
    type: ComponentForwardMsgType,
    data: any
  ): void => {
    if (this.iframeRef.current == null) {
      // This should never happen
      logWarning("Can't send ForwardMsg; missing our iframe!")
      return
    }

    if (this.iframeRef.current.contentWindow == null) {
      // Nor should this!
      logWarning("Can't send ForwardMsg; iframe has no contentWindow!")
      return
    }

    this.iframeRef.current.contentWindow.postMessage(
      {
        isStreamlitMessage: true,
        type: type,
        ...data,
      },
      "*"
    )
  }

  public render = (): ReactNode => {
    const componentId = this.props.element.get("componentId")
    const url = this.props.element.get("url")

    // Determine the component iframe's src. If a URL is specified, we just
    // use that. Otherwise, we derive the URL from the component's ID.
    let src: string
    if (url != null && url !== "") {
      src = url
    } else {
      src = this.props.registry.getComponentURL(componentId, "index.html")
    }

    const renderArgs = JSON.parse(this.props.element.get("argsJson"))
    const renderDfs = this.props.element.get("argsDataframe").toJS()

    if (this.componentReady) {
      // The component has loaded. Send it a new render message immediately.
      this.sendForwardMsg(ComponentForwardMsgType.RENDER, {
        args: renderArgs,
        dfs: renderDfs,
        disabled: this.props.disabled,
      })
    }

    // We always store the most recent render arguments in order to respond
    // to COMPONENT_READY messages. Components can indicate that they're ready
    // multiple times (this will happen if a plugin auto-reloads itself -
    // for example, if it's being served from a webpack dev server). When
    // component sends the COMPONENT_READY message, we send it the most
    // recent render arguments.
    this.lastRenderArgs = renderArgs
    this.lastRenderDataframes = renderDfs

    // Render the iframe. We set scrolling="no", because we don't want
    // scrollbars to appear; instead, we want components to properly auto-size
    // themselves.
    //
    // Without this, there is a potential for a scrollbar to
    // appear for a brief moment after an iframe's content gets bigger,
    // and before it sends the "setFrameHeight" message back to Streamlit.
    //
    // We may ultimately want to give components control over the "scrolling"
    // property.
    //
    // TODO: make sure horizontal scrolling still works!
    return (
      <iframe
        ref={this.iframeRef}
        src={src}
        width={this.props.width}
        height={this.state.frameHeight}
        allowFullScreen={false}
        seamless={true}
        scrolling="no"
        sandbox="allow-forms allow-popups allow-scripts"
      />
    )
  }
}

/** Return the property with the given name, if it exists. */
function tryGetValue(
  obj: any,
  name: string,
  defaultValue: any = undefined
): any {
  return obj.hasOwnProperty(name) ? obj[name] : defaultValue
}
