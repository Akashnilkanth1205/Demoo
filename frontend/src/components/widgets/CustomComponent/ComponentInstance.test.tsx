/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import ErrorElement from "components/shared/ErrorElement"
import { mount, ReactWrapper } from "enzyme"
import { fromJS } from "immutable"
import { buildHttpUri } from "lib/UriUtil"
import { WidgetStateManager } from "lib/WidgetStateManager"
import React from "react"
import { ComponentInstance, Props, State } from "./ComponentInstance"
import { ComponentRegistry } from "./ComponentRegistry"
import { ComponentMessageType, StreamlitMessageType } from "./enums"

// Mock log functions.
jest.mock("lib/log")

// Mock uri utils.
jest.mock("lib/UriUtil")
const mockedBuildHttpUri = buildHttpUri as jest.Mock
mockedBuildHttpUri.mockImplementation(() => "registry/url")

/**
 * Encapsulates all the plumbing for mocking a component,
 * sending it data, and receiving its messages.
 */
class MockComponent {
  public readonly registry: ComponentRegistry

  public readonly wrapper: ReactWrapper<Props, State, ComponentInstance>

  public readonly instance: ComponentInstance

  /**
   * A function to post a mock ComponentMessage from our component iframe
   * to the ComponentInstance under test.
   */
  public readonly sendBackMsg: (type: string, data: any) => void

  /**
   * A mock that will receive ForwardMsgs posted from the ComponentInstance
   * under test to its iframe.
   */
  public readonly receiveForwardMsg: jest.Mock

  public constructor(initialArgs: any = {}, initialDataframes: any = []) {
    const mountNode = document.createElement("div")
    document.body.appendChild(mountNode)

    // mock ComponentRegistry
    this.registry = new ComponentRegistry(() => {
      return {
        host: "streamlit.mock",
        port: 80,
        basePath: "",
      }
    })

    this.registry.registerListener = jest.fn()
    this.registry.deregisterListener = jest.fn()

    // mock Props
    const props = {
      element: fromJS({
        argsDataframe: initialDataframes,
        argsJson: JSON.stringify(initialArgs),
        componentName: "some.component",
        id: "some_id",
        url: "some/url",
      }),
      registry: this.registry,
      width: 100,
      disabled: false,
      widgetMgr: new WidgetStateManager(jest.fn()),
    }

    // Create and mount our ComponentInstance. We need to mount it to an
    // existing DOM element - otherwise, iframe contentWindow is not available.
    this.wrapper = mount(<ComponentInstance {...props} />, {
      attachTo: mountNode,
    })

    this.instance = this.wrapper.instance()

    // Mock the ComponentInstance's iframeRef, and save a reference to
    // its "onBackMsg" function. We use these to simulate data being sent
    // between ComponentInstance and its mounted component.
    const mockIFrameRef = {
      current: { contentWindow: { postMessage: jest.fn() } },
    }
    this.receiveForwardMsg = mockIFrameRef.current.contentWindow.postMessage

    const unsafeInstance = this.instance as any
    this.sendBackMsg = unsafeInstance.onBackMsg
    unsafeInstance.iframeRef = mockIFrameRef
  }
}

describe("ComponentInstance", () => {
  it("should register a message listener on mount", () => {
    const mock = new MockComponent()
    expect(mock.registry.registerListener).toHaveBeenCalled()
    expect(mock.registry.deregisterListener).not.toHaveBeenCalled()
  })

  it("should deregister its message listener on unmount", () => {
    const mock = new MockComponent()
    mock.wrapper.unmount()
    expect(mock.registry.deregisterListener).toHaveBeenCalled()
  })

  describe("COMPONENT_READY handler", () => {
    it("should post a RENDER message to the iframe", () => {
      const jsonArgs = { foo: "string", bar: 5 }

      const mock = new MockComponent(jsonArgs)
      mock.sendBackMsg(ComponentMessageType.COMPONENT_READY, { apiVersion: 1 })
      expect(mock.receiveForwardMsg).toHaveBeenCalledWith(
        renderMsg(jsonArgs, []),
        "*"
      )
    })

    it("should error on unrecognized API version", () => {
      const mock = new MockComponent()
      mock.sendBackMsg(ComponentMessageType.COMPONENT_READY, { apiVersion: 2 })
      const componentError = {
        name: "Error",
        message: "Unrecognized component API version: '2'",
      }
      console.log(mock.wrapper.debug())
      expect(
        mock.wrapper.contains(<ErrorElement {...componentError} />)
      ).toBeTruthy()
    })
  })

  // describe("handleSetComponentValue", () => {
  //   let handleSetComponentValue: (data: any, source: Source) => void
  //   const source = { fromUi: true }
  //
  //   beforeEach(() => {
  //     handleSetComponentValue = instance["handleSetComponentValue"]
  //     WidgetStateManager.prototype.setJsonValue = jest.fn()
  //   })
  //
  //   it("should warn that the `value` prop is missing", () => {
  //     handleSetComponentValue({ value: undefined }, source)
  //     expect(logWarning).toHaveBeenCalledWith(
  //       `handleSetComponentValue: missing 'value' prop`
  //     )
  //     expect(WidgetStateManager.prototype.setJsonValue).not.toHaveBeenCalled()
  //   })
  //
  //   it("should set widget value", () => {
  //     handleSetComponentValue({ value: 1 }, source)
  //     expect(WidgetStateManager.prototype.setJsonValue).toHaveBeenCalledWith(
  //       "some_id",
  //       1,
  //       source
  //     )
  //   })
  // })
  //
  // describe("handleSetFrameHeight", () => {
  //   let handleSetFrameHeight: (data: any) => void
  //
  //   beforeEach(() => {
  //     handleSetFrameHeight = instance["handleSetFrameHeight"]
  //   })
  //
  //   it("should warn that the `height` prop is missing", () => {
  //     handleSetFrameHeight({ height: undefined })
  //     expect(logWarning).toHaveBeenCalledWith(
  //       `handleSetFrameHeight: missing 'height' prop`
  //     )
  //     expect(instance["frameHeight"]).toEqual(0)
  //   })
  //
  //   it("should set iframe height", () => {
  //     handleSetFrameHeight({ height: 100 })
  //     expect(instance["frameHeight"]).toEqual(100)
  //   })
  //
  //   it("should warn that the `iframeRef` prop is missing", () => {
  //     // @ts-ignore
  //     instance["iframeRef"]["current"] = null
  //     handleSetFrameHeight({ height: 200 })
  //     expect(logWarning).toHaveBeenCalledWith(
  //       `handleSetFrameHeight: missing our iframeRef!`
  //     )
  //   })
  // })
  //
  // describe("sendForwardMsg", () => {
  //   let sendForwardMsg: (type: StreamlitMessageType, data: any) => void
  //   const renderType = StreamlitMessageType.RENDER
  //   const data = {
  //     args: { foo: "bar" },
  //     dfs: [],
  //   }
  //
  //   beforeEach(() => {
  //     sendForwardMsg = instance["sendForwardMsg"]
  //   })
  //
  //   it("should call postMessage", () => {
  //     // @ts-ignore
  //     instance["iframeRef"]["current"]["contentWindow"].postMessage = jest.fn()
  //     sendForwardMsg(renderType, data)
  //     expect(
  //       // @ts-ignore
  //       instance["iframeRef"]["current"]["contentWindow"].postMessage
  //     ).toHaveBeenCalledWith({ type: renderType, ...data }, "*")
  //   })
  //
  //   it("should warn that the iframe is missing", () => {
  //     // @ts-ignore
  //     instance["iframeRef"]["current"] = null
  //     sendForwardMsg(renderType, data)
  //     expect(logWarning).toHaveBeenCalledWith(
  //       "Can't send ForwardMsg; missing our iframe!"
  //     )
  //   })
  // })
  //
  // describe("onBackMsg", () => {
  //   let onBackMsg: (type: string, data: any) => void
  //
  //   beforeEach(() => {
  //     onBackMsg = instance["onBackMsg"]
  //   })
  //
  //   describe("COMPONENT_READY", () => {
  //     const componentReadyType = ComponentMessageType.COMPONENT_READY
  //
  //     it("should throw `Unrecognized component API version` exception", () => {
  //       onBackMsg(componentReadyType, { apiVersion: "bad_api_version" })
  //       // @ts-ignore
  //       expect(instance["state"]["componentError"]["message"]).toBe(
  //         "Unrecognized component API version: 'bad_api_version'"
  //       )
  //     })
  //
  //     it("should set component ready to true", () => {
  //       onBackMsg(componentReadyType, { apiVersion: 1 })
  //       expect(instance["componentReady"]).toBe(true)
  //     })
  //
  //     it("should call postMessage", () => {
  //       instance["sendForwardMsg"] = jest.fn()
  //       onBackMsg(componentReadyType, { apiVersion: 1 })
  //       expect(instance["sendForwardMsg"]).toHaveBeenCalledWith(
  //         StreamlitMessageType.RENDER,
  //         {
  //           args: { foo: "bar" },
  //           dfs: [],
  //         }
  //       )
  //     })
  //   })
  //
  //   describe("SET_COMPONENT_VALUE", () => {
  //     const type = ComponentMessageType.SET_COMPONENT_VALUE
  //     const data = { value: 1 }
  //
  //     it("should call handleSetComponentValue", () => {
  //       instance["handleSetComponentValue"] = jest.fn()
  //       instance["componentReady"] = true
  //       onBackMsg(type, data)
  //       expect(instance["handleSetComponentValue"]).toHaveBeenCalledWith(
  //         data,
  //         {
  //           fromUi: true,
  //         }
  //       )
  //     })
  //
  //     it("should warn component is not ready", () => {
  //       instance["componentReady"] = false
  //       onBackMsg(type, data)
  //       expect(logWarning).toHaveBeenCalledWith(
  //         "Got streamlit:setComponentValue before streamlit:componentReady!"
  //       )
  //     })
  //   })
  //
  //   describe("SET_FRAME_HEIGHT", () => {
  //     const setFrameHeightType = ComponentMessageType.SET_FRAME_HEIGHT
  //     const data = { height: 100 }
  //
  //     it("should call handleSetFrameHeight", () => {
  //       instance["handleSetFrameHeight"] = jest.fn()
  //       instance["componentReady"] = true
  //       onBackMsg(setFrameHeightType, data)
  //       expect(instance["handleSetFrameHeight"]).toHaveBeenCalledWith(data)
  //     })
  //
  //     it("should warn that the component is not ready", () => {
  //       instance["componentReady"] = false
  //       onBackMsg(setFrameHeightType, data)
  //       expect(logWarning).toHaveBeenCalledWith(
  //         "Got streamlit:setFrameHeight before streamlit:componentReady!"
  //       )
  //     })
  //   })
  //
  //   it("should warn with `Unrecognized ComponentBackMsgType`", () => {
  //     const type = "Wrong type"
  //     onBackMsg(type, {})
  //     expect(logWarning).toHaveBeenCalledWith(
  //       "Unrecognized ComponentBackMsgType: Wrong type"
  //     )
  //   })
  // })
  //
  // describe("Render iframe", () => {
  //   const element = document.createElement("div")
  //   document.body.appendChild(element)
  //
  //   it("should render ErrorElement", () => {
  //     const componentError = {
  //       name: "foo",
  //       message: "bar",
  //     }
  //     wrapper.setState({ componentError })
  //     expect(
  //       wrapper.containsMatchingElement(<ErrorElement {...componentError} />)
  //     ).toBeTruthy()
  //   })
  //
  //   it("should render iframe and set `src` to registry/url", () => {
  //     const props = getProps({ url: "" })
  //     const wrapper = mount<ComponentInstance>(
  //       <ComponentInstance {...props} />,
  //       {
  //         attachTo: element,
  //       }
  //     )
  //     expect(wrapper.find("iframe").props()).toMatchSnapshot()
  //   })
  //
  //   it("should render iframe and set `src` to element url", () => {
  //     expect(wrapper.find("iframe").props()).toMatchSnapshot()
  //   })
  // })
})

function renderMsg(args: any, dataframes: any, disabled?: boolean): any {
  return forwardMsg(StreamlitMessageType.RENDER, {
    args,
    dfs: dataframes,
    disabled,
  })
}

function forwardMsg(type: StreamlitMessageType, data: any): any {
  return { type, ...data }
}
