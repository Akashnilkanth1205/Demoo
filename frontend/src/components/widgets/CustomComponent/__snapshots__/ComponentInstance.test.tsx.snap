// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentInstance Render iframe should render iframe and set \`src\` to element url 1`] = `
Object {
  "allow": "accelerometer; ambient-light-sensor; autoplay; battery; camera; document-domain; encrypted-media; geolocation; gyroscope; layout-animations; legacy-image-formats; magnetometer; microphone; midi; oversized-images; payment; picture-in-picture; publickey-credentials-get; sync-xhr; usb; vr ; wake-lock; xr-spatial-tracking",
  "height": 0,
  "sandbox": "allow-forms allow-modals allow-popups allow-popups-to-escape-sandbox allow-scripts",
  "scrolling": "no",
  "src": "some/url?streamlitUrl=http%3A%2F%2Flocalhost%2F",
  "title": "some.component",
  "width": 100,
}
`;

exports[`ComponentInstance Render iframe should render iframe and set \`src\` to registry/url 1`] = `
Object {
  "allow": "accelerometer; ambient-light-sensor; autoplay; battery; camera; document-domain; encrypted-media; geolocation; gyroscope; layout-animations; legacy-image-formats; magnetometer; microphone; midi; oversized-images; payment; picture-in-picture; publickey-credentials-get; sync-xhr; usb; vr ; wake-lock; xr-spatial-tracking",
  "height": 0,
  "sandbox": "allow-forms allow-modals allow-popups allow-popups-to-escape-sandbox allow-scripts",
  "scrolling": "no",
  "src": "registry/url?streamlitUrl=http%3A%2F%2Flocalhost%2F",
  "title": "some.component",
  "width": 100,
}
`;
